### PR TITLE
Fix memory management errors caused by not retaining the arguments on an...

### PR DESCRIPTION
**Pull request by rhgills**
... NSInvocation that we're caching.

### CODE REVIEW

**Comment by rhgills** [comment]
r @kyleve @alanf

File: Source/OCMock/OCMockObject.m [hunk 1]
```diff
@@ -225,6 +225,7 @@ - (BOOL)handleInvocation:(NSInvocation *)anInvocation
 	OCMockRecorder *recorder = nil;
 	unsigned int			   i;
 
+    [anInvocation retainArguments];
```
Comment (kyleve) [inline Source/OCMock/OCMockObject.m]: Do we need to release them too?

Also, tests pls

**Comment by rhgills** [comment]
Background: was causing memory management errors, as OCMock was, in order to verify expectations, pulling arguments from an NSInvocation (that was called on the mock object), but those arguments had already been deallocated.

File: Source/OCMock/OCMockObject.m (hunk 1 above)
Comment (rhgills) [inline Source/OCMock/OCMockObject.m]: NSInvocation should be responsible for releasing them when it is being deallocated, but good point, let me verify this.

Will also add a test.

**Comment by rhgills** [comment]
PR updated, please take another look @kyleve and @alanf!

File: Source/OCMock/NSInvocation+OCMAdditions.m [hunk 2]
```diff
@@ -9,6 +9,16 @@
...
 @implementation NSInvocation(OCMAdditions)
 
+- (BOOL)isArgumentAtIndexAnObject:(int)argIndex
+{
+    const char *argType = OCMTypeWithoutQualifiers([[self methodSignature] getArgumentTypeAtIndex:argIndex]);
+
+    if((strlen(argType) > 1) && (strchr("{^", argType[0]) == NULL) && (strcmp("@?", argType) != 0))
+        [NSException raise:NSInvalidArgumentException format:@"Cannot handle argument type '%s'.", argType];
+
+    return argType[0] == '@' || argType[0] == '#';
```
Comment (kyleve) [inline Source/OCMock/NSInvocation+OCMAdditions.m]: What's the `#`?

File: Source/OCMock/OCMockObject.m [hunk 3]
```diff
@@ -328,5 +344,23 @@ - (NSString *)_recorderDescriptions:(BOOL)onlyExpectations
 	return outputString;
 }
 
+typedef void (^OCMockInvocationObjectArgumentBlock)(id object);
+- (void)_performBlock:(OCMockInvocationObjectArgumentBlock)block onObjectArgumentsOfInvocation:(NSInvocation *)anInvocation;
```
Comment (kyleve) [inline Source/OCMock/OCMockObject.m]: `_enumerateObjectArgumentsOfInvocation:usingBlock:`

File: Source/OCMock/OCMockObject.m [hunk 4]
```diff
@@ -328,5 +344,23 @@ - (NSString *)_recorderDescriptions:(BOOL)onlyExpectations
...
+    NSUInteger numberOfArguments = anInvocation.methodSignature.numberOfArguments;
+
+    // Self and _cmd are the first two arguments, which we don't care about.
+    // Start at 2.
+    for (NSUInteger i = 2; i < numberOfArguments; i++) {
+        if (![anInvocation isArgumentAtIndexAnObject:i]) {
+            continue;
+        }
+
+        id argument;
```
Comment (kyleve) [inline Source/OCMock/OCMockObject.m]: assign to nil (OCMock isn't using ARC)

File: Source/OCMockTests/OCMockObjectTests.m [hunk 5]
```diff
@@ -81,7 +81,8 @@ - (void)receiveNotification:(NSNotification *)aNotification
```
Comment (kyleve) [inline Source/OCMockTests/OCMockObjectTests.m]: I assume you saw these tests fail before making your changes?

File: Source/OCMock/NSInvocation+OCMAdditions.h [hunk 6]
```diff
@@ -9,6 +9,8 @@
 
 - (id)getArgumentAtIndexAsObject:(int)argIndex;
 
+- (BOOL)isArgumentAtIndexAnObject:(int)argIndex;
```
Comment (alanf) [inline Source/OCMock/NSInvocation+OCMAdditions.h]: `argumentAtIndexIsObject:` might make this method easier to distinguish from the very similar looking `getArgumentAtIndexAsObject` (I had to read it a couple of times to pick up on `As` v. `An`)

**Comment by alanf** [comment]
LGTM

File: Source/OCMock/NSInvocation+OCMAdditions.m (hunk 2 above)
Comment (rhgills) [inline Source/OCMock/NSInvocation+OCMAdditions.m]: From https://developer.apple.com/library/mac/documentation/Cocoa/Conceptual/ObjCRuntimeGuide/Articles/ocrtTypeEncodings.html, 

`#` is a class object (`Class`)

File: Source/OCMockTests/OCMockObjectTests.m (hunk 5 above)
Comment (rhgills) [inline Source/OCMockTests/OCMockObjectTests.m]: Yup.

File: Source/OCMock/OCMockObject.m (hunk 4 above)
Comment (rhgills) [inline Source/OCMock/OCMockObject.m]: Will do... curious though, is this more than a stylistic matter pre-ARC?

File: Source/OCMock/NSInvocation+OCMAdditions.h (hunk 6 above)
Comment (rhgills) [inline Source/OCMock/NSInvocation+OCMAdditions.h]: Thats actually the name I started with. Changed to `isArgumentAtIndexAnObject` to have the `is`, indicating a method returning a `BOOL`, at the start of the method. Will change though, I agree the methods are very similar.

File: Source/OCMock/OCMockObject.m (hunk 3 above)
Comment (rhgills) [inline Source/OCMock/OCMockObject.m]: :+1:

**Comment by rhgills** [comment]
@alanf and @kyleve, pushed changes addressing your feedback. Please take another look!

File: Source/OCMock/NSInvocation+OCMAdditions.m (hunk 2 above)
Comment (kyleve) [inline Source/OCMock/NSInvocation+OCMAdditions.m]: And here I thought it was the hashtag object

File: Source/OCMockTests/OCMockObjectTests.m [hunk 7]
```diff
@@ -89,21 +90,46 @@ @interface OCMockObjectTests : XCTestCase
...
+    [super setUp];
+
 	mock = [OCMockObject mockForClass:[NSString class]];
+
+    self.deallocatedTestObjectPointers = [[NSMutableSet alloc] init];
 }
 
+- (void)tearDown;
+{
+    [self.deallocatedTestObjectPointers release];
```
Comment (kyleve) [inline Source/OCMockTests/OCMockObjectTests.m]: nil this out?

**Comment by rhgills** [comment]
Declining this in favor of private fork on Stash. Please continue reviewing there!

**Comment by kyleve** [comment]
arugh

File: Source/OCMockTests/OCMockObjectTests.m (hunk 7 above)
Comment (rhgills) [inline Source/OCMockTests/OCMockObjectTests.m]: Ahh yes! Comment was misplaced before, I think

**Comment by rhgills** [comment]
Sorry buddy

**Pull request closed without merging**